### PR TITLE
Speed up fetching device lists changes in sync.

### PR DESCRIPTION
**Pull request by erikjohnston**
Currently we copy `users_who_share_room` needlessly about three times,
which is expensive when the set is large (which it can easily be).

This is based off a commit that is common to `develop` and `matrix-org-hotfixes`

### CODE REVIEW

File: synapse/util/caches/stream_change_cache.py [hunk 1]
```diff
@@ -85,7 +86,7 @@ def has_entity_changed(self, entity: EntityType, stream_pos: int) -> bool:
         return False
 
     def get_entities_changed(
-        self, entities: Iterable[EntityType], stream_pos: int
+        self, entities: Collection[EntityType], stream_pos: int
```
Comment (erikjohnston) [inline synapse/util/caches/stream_change_cache.py:89]: I've checked all the callers and this matches their types fwiw

File: synapse/util/caches/stream_change_cache.py [hunk 2]
```diff
@@ -94,7 +95,17 @@ def get_entities_changed(
         """
         changed_entities = self.get_all_entities_changed(stream_pos)
         if changed_entities is not None:
-            result = set(changed_entities).intersection(entities)
+            # We now do an intersection, trying to do so in the most efficient
+            # way possible (some of these sets are *large*). First check in the
+            # given iterable is already set that we can reuse, otherwise we
+            # create a set of the *smallest* of the two iterables and call
+            # `intersection(..)` on it (this can be twice as fast as the reverse).
+            if isinstance(changed_entities, (set, frozenset)):
```
Comment (clokep) [inline synapse/util/caches/stream_change_cache.py:103]: Is this meant to be `entities` here? I think `changed_entities` will always be ` list` (`get_all_entities_changed` returns `None` or a `list`).

Comment (erikjohnston) [reply synapse/util/caches/stream_change_cache.py]: whoops, yes

File: synapse/util/caches/stream_change_cache.py [hunk 3]
```diff
@@ -94,7 +95,17 @@ def get_entities_changed(
...
         changed_entities = self.get_all_entities_changed(stream_pos)
         if changed_entities is not None:
-            result = set(changed_entities).intersection(entities)
+            # We now do an intersection, trying to do so in the most efficient
+            # way possible (some of these sets are *large*). First check in the
+            # given iterable is already set that we can reuse, otherwise we
+            # create a set of the *smallest* of the two iterables and call
+            # `intersection(..)` on it (this can be twice as fast as the reverse).
+            if isinstance(entities, (set, frozenset)):
+                result = changed_entities.intersection(entities)
```
Comment (clokep) [inline synapse/util/caches/stream_change_cache.py:104]:
```suggestion
                result = entities.intersection(changed_entities)
```

Comment (erikjohnston) [reply synapse/util/caches/stream_change_cache.py]: :man_facepalming:

File: synapse/util/caches/stream_change_cache.py [hunk 4]
```diff
@@ -94,7 +95,17 @@ def get_entities_changed(
         """
         changed_entities = self.get_all_entities_changed(stream_pos)
         if changed_entities is not None:
-            result = set(changed_entities).intersection(entities)
+            # We now do an intersection, trying to do so in the most efficient
+            # way possible (some of these sets are *large*). First check in the
+            # given iterable is already set that we can reuse, otherwise we
+            # create a set of the *smallest* of the two iterables and call
+            # `intersection(..)` on it (this can be twice as fast as the reverse).
```
Comment (clokep) [inline synapse/util/caches/stream_change_cache.py:102]: https://wiki.python.org/moin/TimeComplexity#set has some stuff to say about this:
> Average case: O(min(len(s), len(t)) 
> Worst case: O(len(s) * len(t))
> 
> replace "min" with "max" if t is not a set

Comment (erikjohnston) [reply synapse/util/caches/stream_change_cache.py:102]: Well that is interesting, I did:

```
In [1]: from synapse.util.stringutils import random_string                                                                                                                                                         

In [2]:                                                                                                                                                                                                            

In [2]: a = [random_string(10) for _ in range(10000)]                                                                                                                                                              

In [3]: b = [random_string(10) for _ in range(10)]                                                                                                                                                                 

In [4]: %timeit set(a).intersection(b)                                                                                                                                                                             
299 µs ± 3.41 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [5]: %timeit set(b).intersection(a)                                                                                                                                                                             
141 µs ± 9.19 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

I'm slightly surprised that they're symmetric when `t` is not a set

(I'm going to wait for you to point out my embarrassing typo now....)

Comment (clokep) [reply synapse/util/caches/stream_change_cache.py:102]: This looks sane to me. The only thing I can think of is that this is partially bound by the creation of the `set()` and not by the `intersction` code.

Comment (erikjohnston) [reply synapse/util/caches/stream_change_cache.py:102]: Ah, that'd make sense if they ignored the time requirements for creating the set `s`

Comment (clokep) [reply synapse/util/caches/stream_change_cache.py:102]: Right, so it'd be interesting to see:

```pycon
>>> sa = set(a)
>>> sb = set(b)
>>> %timeit sa.interesection(b)
>>> %timeit sb.interesection(a)
```

File: changelog.d/7423.misc [hunk 5]
```diff
@@ -0,0 +1 @@
+Speed up fetching device lists changes in sync.
```
Comment (richvdh) [inline changelog.d/7423.misc:1]:
```suggestion
Speed up fetching device lists changes when handling `/sync` requests.
```

File: synapse/util/caches/stream_change_cache.py (hunk 4 above)
Comment (clokep) [reply synapse/util/caches/stream_change_cache.py:102]: I looked at this a bit more and I think your conclusion is sane, but the gist of my exploration is that converting things to sets is expensive. 👍

**Pull request merged**